Antistasi-Ultimate-Community/A3-Antistasi-Ultimate
# Add support for weighted-list faction templates.

## What type of PR is this?
1. [ ] Bug
2. [ ] Change
3. [x] Enhancement
4. [ ] Miscellaneous

### What have you changed and why?
Firstly, this adds support for constructing faction templates using weighted lists. See [https://community.bistudio.com/wiki/selectRandomWeighted](https://community.bistudio.com/wiki/selectRandomWeighted) for details. This allows template makers to create factions that spawn with an uneven distribution of weapons without making a large number of duplicate entries in the template, which is especially fun for more 'irregular' opponents such as SF and rivals. It also helps gameplay by giving factions common weapons that players can unlock easily, and rare weapons that can be a fun loot drop. This supports everything individual soldiers in a faction can spawn with, but not vehicles as implementing the same for vehicles would involve changes in a lot of different files.

In addition, this adds support for randomly choosing attachments (optics, suppressors, etc) in weapons by detecting if the weapon has a classname or an array, which simplifies faction templates by removing the number of permutations of a given weapon entry needed to have a variety of attachments - especially if e.g. multiple types of optic, accessory, etc all get mixed together. This functionality also supports weighted lists as well as regular arrays of classnames.

### Please specify which Issue this PR Resolves (If Applicable).
(N/A)

### Please verify the following.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?

1. [x] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: Create or modify a faction to implements this feature (e.g. by adding weights to their weapon spawn lists, putting arrays in the attachment slots of their weapons, etc). An example of what this looks like in practice is below since it's easier than explaining in words.
![image](https://github.com/user-attachments/assets/ade78de6-4a9a-48e5-be14-c9b360bed216)
In this example, Mk20 rifles spawn at a 3:1 rate compared to TRG-21s, regular soldiers have ACO optics at a 1:4 rate compared to ironsights, team leaders have an equal distribution of ACOs, MRCOs and ironsights. When dealing with DLC compatibility, take care to make sure to keep track of item weights - if a spawn table is weighted, make sure to `append [item, weight]` instead of just `pushBack item`.
Once this is done, load the template in question in to a new Antistasi game and check the inventories of spawned soldiers.
********************************************************
Notes: I am currently working on modifying the vanilla AAF template to provide a thorough example of this implemented, but as that takes a long while I'm putting up this PR now.
